test/distributed/fsdp/test_fsdp_grad_acc.py

```diff
@@ -0,0 +1,263 @@
+# Owner(s): ["oncall: distributed"]
+
+import contextlib
+import itertools
+import sys
+from typing import List, Optional, Tuple
+from dataclasses import dataclass
+
+import torch
+from torch import distributed as dist
+from torch.distributed.fsdp import CPUOffload
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp.fully_sharded_data_parallel import BackwardPrefetch
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import FSDPTest
+from torch.testing._internal.common_utils import (
+    TEST_WITH_DEV_DBG_ASAN,
+    instantiate_parametrized_tests,
+    parametrize,
+    run_tests,
+)
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+@dataclass
+class _GradAccConfig:
+    """
+    This configures how gradients are accumulated in :meth:`_test_grad_acc`.
+    Each instance of this class represents ``num_iters``-many consecutive
+    iterations, where the ``no_sync()`` context manager is used or not as given
+    by ``use_no_sync``.
+
+    Attributes:
+        use_no_sync (bool): Indicates whether to use the ``no_sync()`` context
+            manager as the way to accumulate gradients.
+        num_iters (int): Number of iterations to accumulate gradients.
+    """
+    use_no_sync: bool
+    num_iters: int
+
+    def __repr__(self) -> str:
+        # Override to remove any spaces in the string to appease the internal
+        # build's test name parser
+        return (
+            f"(use_no_sync={self.use_no_sync},"
+            f"num_iters={self.num_iters})"
+        )
+
+
+@dataclass
+class _GradAccConfigs:
+    """
+    This wraps a :class:`list` of :class:`_GradAccConfig` instances with the
+    sole purpose of overriding :meth:`__repr__` to remove spaces.
+    """
+    configs: List[_GradAccConfig]
+
+    def __repr__(self) -> str:
+        # Override to remove any spaces in the string to appease the internal
+        # build's test name parser
+        return (
+            "[" + ",".join(config.__repr__() for config in self.configs) + "]"
+        )
+
+
+class TestGradAcc(FSDPTest):
```

Comment:
Can we get a general idea of how much these tests increase TTS for OSS CI? For example I wonder if it is costly enough, should we consider running some of these tests on master-only.

Reply:
Updated version with only the two interleaved tests takes ~60 seconds on the AWS cluster.